src/connectors/exchangeRate/index.ts

```diff
@@ -0,0 +1,251 @@
+import axios from 'axios'
+
+import { CACHE_TTL } from 'common/enums'
+import { environment } from 'common/environment'
+import { NetworkError, UnknownError } from 'common/errors'
+import logger from 'common/logger'
+import { CacheService } from 'connectors'
+import {
+  GQLExchangeRate,
+  GQLQuoteCurrency,
+  GQLTransactionCurrency,
+} from 'definitions'
+
+// TYPES
+
+interface Pair {
+  from: GQLTransactionCurrency
+  to: GQLQuoteCurrency
+}
+
+type TokenCurrency = GQLTransactionCurrency.LIKE | GQLTransactionCurrency.USDT
+type FiatCurrency = GQLTransactionCurrency.HKD
+
+// CONSTANTS
+
+const tokenCurrencies: TokenCurrency[] = [
+  'LIKE' as GQLTransactionCurrency.LIKE,
+  'USDT' as GQLTransactionCurrency.USDT,
+]
+const fiatCurrencies: FiatCurrency[] = ['HKD' as GQLTransactionCurrency.HKD]
+const quoteCurrencies: GQLQuoteCurrency[] = [
+  'TWD' as GQLQuoteCurrency.TWD,
+  'HKD' as GQLQuoteCurrency.HKD,
+  'USD' as GQLQuoteCurrency.USD,
+]
+
+const TOKEN_TO_COINGECKO_ID = {
+  LIKE: 'likecoin',
+  USDT: 'tether',
+} as const
+
+const COINGECKO_API_URL = 'https://api.coingecko.com/api/v3/simple/price'
+const EXCHANGE_RATES_DATA_API_URL =
+  'https://api.apilayer.com/exchangerates_data/latest'
+
+// MAIN
+
+export class ExchangeRate {
+  cache: CacheService
+  expire: number
+  constructor() {
+    this.cache = new CacheService('exchangeRate')
+    this.expire = CACHE_TTL.STATIC
+  }
+
+  getRates = async (
+    from?: GQLTransactionCurrency,
+    to?: GQLQuoteCurrency
+  ): Promise<GQLExchangeRate[] | never> => {
+    const allPairs = [...this.getTokenPairs(), ...this.getFiatPairs()]
+    let pairs = allPairs
+    if (from) {
+      pairs = pairs.filter((p) => p.from === from)
+    }
+    if (to) {
+      pairs = pairs.filter((p) => p.to === to)
+    }
+
+    return Promise.all(pairs.map((p) => this.getRate(p)))
+  }
+
+  updateTokenRates = async () => {
+    await this.updateRatesToCache(await this.fetchTokenRates())
+  }
+
+  updateFiatRates = async () => {
+    await this.updateRatesToCache(await this.fetchFiatRates())
+  }
+
+  private getRate = async (pair: Pair): Promise<GQLExchangeRate | never> => {
+    const data = (await this.cache.getObject({
+      keys: this.genCacheKeys(pair),
+      getter: async () => this.updateAllRatesAndGetRate(pair),
```

Comment:
Why don't just get the rate of this `pair` instead since there is concurrent issue if `getRates`'s `from` and `to` are `null` (means `getRate` of all pairs)?

Reply:
Yes, concurrent will be a issue. I will fix it